crates/libs/core/src/future.rs

```diff
@@ -0,0 +1,74 @@
+#![cfg(feature = "std")]
+
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Poll, Waker},
+};
+
+/// Wraps an `IAsyncOperation`, `IAsyncOperationWithProgress`, `IAsyncAction`, or `IAsyncActionWithProgress`.
+/// Impls for this trait are generated automatically by windows-bindgen.
+pub trait AsyncOperation {
+    /// The type produced when the operation finishes.
+    type Output;
+    /// Returns whether the operation is finished, in which case `self.get_results()` can be used to get the returned data.
+    /// Wraps `self.Status() != AsyncStatus::Started`.
+    fn is_complete(&self) -> crate::Result<bool>;
+    /// Register a callback that will be called once the operation is finished.
+    /// This can only be called once.
+    /// Wraps `self.SetCompleted(f)`.
+    fn set_completed(&self, f: impl Fn() + Send + 'static) -> crate::Result<()>;
+    /// Get the result value from a completed operation.
+    /// Wraps `self.GetResults()`.
+    fn get_results(&self) -> crate::Result<Self::Output>;
+    /// Attempts to cancel the operation. Any error is ignored.
+    /// Wraps `self.Cancel()`.
+    fn cancel(&self);
+}
+
+/// A wrapper around an `AsyncOperation` that implements `std::future::Future`.
+/// This is used by generated `IntoFuture` impls. It shouldn't be necessary to use this type manually.
+pub struct FutureWrapper<T: AsyncOperation> {
+    inner: T,
+    waker: Option<Arc<Mutex<Waker>>>,
```

Comment:
This seems like an overly expensive combination for synchronization. I'm not sure how to simplify but I'll play around with it a bit.

Reply:
Haven't found a simpler approach. 🤷

Reply:
How does this compare to the old implementation? Is this incurring an extra allocation for each Future call? :o

Reply:
Yes, previously (#3142) I didn't store the `Waker` at all. Now the `Waker` is stored in an `Arc`, which adds a heap allocation, and requires a bunch of interlocking to retrieve the `Waker`.

Reply:
This is why I'm not excited about this in general - the WinRT async model and the Rust futures library don't seem to have a natural fit.

Reply:
I don't feel like an extra heap allocation is a huge problem since eg. calling SetCompleted also needs to allocate already.
I think it is possible to combine those into one allocation though at the cost of some unsafe code. Would you prefer that?
Also, would you prefer to replace the Mutex with lighter weight synchronization using hand-rolled atomics?

Reply:
I'd prefer to keep it as simple as possible. We can always optimize later as needed.

Reply:
Perhaps at least a comment here somewhere that the `Completed` can only be set once and thus the need for a shared waker.